src/cpu/simple/probes/looppoint_analysis.cc

```diff
@@ -0,0 +1,280 @@
+                params.bb_excluded_addr_ranges[i].end()
+            )
+        );
+        DPRINTF(LooppointAnalysis, "Excluding address range: (%li, %li)\n",
+                params.bb_excluded_addr_ranges[i].start(),
+                params.bb_excluded_addr_ranges[i].end()
+        );
+    }
+
+    DPRINTF(LooppointAnalysis, "Valid address range: (%li, %li)\n",
+            bbValidAddrRange.start(), bbValidAddrRange.end());
+    DPRINTF(LooppointAnalysis, "Valid marker address range: (%li, %li)\n",
+            markerValidAddrRange.start(), markerValidAddrRange.end());
+}
+
+void
+LooppointAnalysis::updateLocalBBV(const Addr pc)
+{
+    if (localBBV.find(pc) == localBBV.end())
+    {
+        localBBV.insert(std::make_pair(pc, 1));
+    }
+    else
+    {
+        localBBV.find(pc)->second++;
+    }
+}
+
+void
+LooppointAnalysis::checkPc(const std::pair<SimpleThread*,
+                                                     StaticInstPtr>& inst_pair)
+{
+    const StaticInstPtr &inst = inst_pair.second;
+
+    if (inst->isMicroop() && !inst->isLastMicroop())
+    {
+        // ignore this if it is a microop
+        return;
+    }
+
+    SimpleThread* thread = inst_pair.first;
+
+    auto &pcstate =
+                thread->getTC()->pcState().as<GenericISA::PCStateWithNext>();
+    Addr pc = pcstate.pc();
+
+    if (lpaManager->ifEncountered(pc))
+    {
+        // if we have already encountered this pc, we should already
+        // categorized it
+        if (lpaManager->ifValidNotControl(pc))
+        {
+            // if it is categorized as a valid not control instruction
+            bbInstCounter++;
+            lpaManager->incrementGlobalInstCounter();
+        }
+        else if (lpaManager->ifValidControl(pc))
+        {
+            // if it is categorized as a valid control instruction
+            bbInstCounter ++;
+            lpaManager->incrementGlobalInstCounter();
+            lpaManager->updateBBInstMap(pc, bbInstCounter);
+            updateLocalBBV(pc);
+            lpaManager->updateGlobalBBV(pc);
+            bbInstCounter = 0;
+            if (lpaManager->ifBackwardBranch(pc))
+            {
+                // if it is categorized as a backward branch
+                lpaManager->countBackwardBranch(pc);
+            }
+        }
+        return;
+    }
+
+    lpaManager->updateEncountered(pc);
+    // if we have not encountered this pc before, we should now update it to
+    // the corresponding category
+
+    if (!thread->getIsaPtr()->inUserMode())
+    {
+        // ignore this if it is not in user mode
+        return;
+    }
+
+    if (bbValidAddrRange.end() > 0 &&
+        (pc < bbValidAddrRange.start() || pc > bbValidAddrRange.end()))
+    {
+        // ignore this if it is not in the valid address range
+        return;
+    }
+
+    if (bbExcludedAddrRanges.size() > 0)
+    {
+        for (int i = 0; i < bbExcludedAddrRanges.size(); i++)
+        {
+            if (pc >= bbExcludedAddrRanges[i].start() &&
```

Comment:
Can you use `contains` (https://github.com/gem5/gem5/blob/stable/src/base/addr_range.hh#L489)?